core/securitykey.py

```diff
@@ -26,88 +26,99 @@
 
         Therefor that header is prefixed with "Sec-" - so it cannot be read or set using JavaScript.
 """
-from viur.core import utils, current, db, tasks
+import typing
+import datetime
+from viur.core import conf, utils, current, db, tasks
 from viur.core.tasks import DeleteEntitiesIter
-from datetime import datetime, timedelta
-from typing import Union
 
 SECURITYKEY_KINDNAME = "viur-securitykey"
+SECURITYKEY_DURATION = 24 * 60 * 60  # one day
 
 
-def create(duration: Union[None, int] = None, **custom_data) -> str:
+def create(duration: typing.Union[None, int] = None, session: bool = True, **custom_data) -> str:
     """
-        Creates a new one-time security key or returns the current sessions CSRF-token.
+        Creates a new one-time security key or returns a valid CSRF-token for the current session.
 
         The custom data (given as **custom_data) that can be stored with the key.
         Any data provided must be serializable by the datastore.
 
-        :param duration: Make this key valid for a fixed timeframe of seconds (and independent of the current session)
+        :param duration: Make this key valid for a fixed timeframe of seconds
+        :param session: Bind this key to the current session
+        :param custom_data: Any other data is stored behind the skey.
+
         :returns: The new one-time key, which is a randomized string.
     """
-    if not duration:
-        if custom_data:
-            raise ValueError("Providing any custom_data is not allowed when session security key is wanted")
+    if any(k.startswith("viur_") for k in custom_data):
+        raise ValueError("custom_data keys with a 'viur_'-prefix are reserved.")
 
-        return current.session.get().getSecurityKey()
-
-    elif "until" in custom_data:
-        raise ValueError("The 'until' property is reserved and cannot be used in the custom_data")
+    if not duration:
+        duration = conf["viur.session.lifeTime"] if session else SECURITYKEY_DURATION
 
     key = utils.generateRandomString()
 
     entity = db.Entity(db.Key(SECURITYKEY_KINDNAME, key))
     entity |= custom_data
 
-    entity["until"] = utils.utcNow() + timedelta(seconds=int(duration))
+    entity["viur_session"] = current.session.get().cookie_key if session else None
+    entity["viur_until"] = utils.utcNow() + datetime.timedelta(seconds=int(duration))
     db.Put(entity)
 
     return key
 
 
-def validate(key: str, useSessionKey: bool) -> Union[bool, db.Entity]:
+def validate(key: str, session: bool = True) -> typing.Union[bool, db.Entity]:
```

Comment:
I would like to have a more expressive name in the sense of "enforce that skey must be bound to the current session" or "only accept skeys from this session", but I have no really good suggestion for a short parameter name :(




```suggestion
def validate(key: str, enforce_session_bounded: bool = True) -> typing.Union[bool, db.Entity]:
```

Reply:
I've first named this `session_bounded`. We should discuss this on Friday as well.

Reply:
It will be renamed to `session_bounded`.

Reply:
[3340b47](https://github.com/viur-framework/viur-core/pull/764/commits/3340b474e64c07dc1ab7e5e686e48229b1c1ac03) renames the parameter into `session_bound`.